bokehjs/src/lib/models/glyphs/webgl/annulus.ts

```diff
@@ -0,0 +1,34 @@
+import {ReglWrapper} from "./regl_wrap"
+import {Float32Buffer} from "./buffer"
+import {SXSYGlyphGL} from "./sxsy"
+import {GLMarkerType} from "./types"
+import type {AnnulusView} from "../annulus"
+
+export class AnnulusGL extends SXSYGlyphGL {
+  constructor(regl_wrapper: ReglWrapper, override readonly glyph: AnnulusView) {
+    super(regl_wrapper, glyph)
+  }
+
+  get marker_type(): GLMarkerType {
+    return "annulus"
+  }
+
+  get outer_radius(): Float32Buffer {
+    return this._widths
+  }
+
+  get inner_radius(): Float32Buffer {
+    return this._heights
+  }
+
+  protected override _set_data(): void {
+    super._set_data()
+
+    this.outer_radius.set_from_array(this.glyph.souter_radius)
+    this.inner_radius.set_from_array(this.glyph.sinner_radius)
+
+    // unused
+    this._angles.set_from_scalar(0, 1)
```

Comment:
```suggestion
    this._angles.set_from_scalar(0)
```
We should leave the second argument here to be the default of `4`. There are only two options for the length of such buffers, either one per marker rendered or one for each of the 4 vertices of the instanced rectangle if we want every marker to have the same value. Sending a buffer of length 1 is potentially dangerous, even if it is not used. Evidently it works fine as it is on my and your test systems, but it could well break on some arbitrary graphics card that implements its CPU to GPU memory transfer in a different way. So it is safest to leave it as 4.

Also, as `this._angles` is unused we only really need to set it the first time we call `_set_data()`. Resetting it to the same values will cause it to be send to the GPU again, which will make it slightly slower than necessary although it may be unnoticeable in the real world. Maybe it could be wrapped in an `if`-statement that checks if the `WrappedBuffer.array` is set or not? This isn't currently exposed by `WrappedBuffer` but some sort of "are you initialised" function could be added.

Reply:
There are a few other places that use the same code.

Reply:
I misunderstood the meaning of this length argument. I reverted all to the default.

Reply:
I added `_set_once()` for those unused attributes, and new `set_data()` with now manages both `_set_data()` and `_set_once()`.